File: src/main/java/com/gmail/borlandlp/minigamesdtools/listener/Events.java
package com.gmail.borlandlp.minigamesdtools.listener;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools;
import com.gmail.borlandlp.minigamesdtools.arena.ArenaPlayersRelative;
import com.gmail.borlandlp.minigamesdtools.arena.localevent.*;
import com.gmail.borlandlp.minigamesdtools.arena.ArenaBase;
import com.gmail.borlandlp.minigamesdtools.events.ArenaGameEndedEvent;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.*;
import org.bukkit.event.player.*;
import org.bukkit.event.weather.WeatherChangeEvent;
import org.bukkit.inventory.ItemStack;

public class Events implements Listener {

    @EventHandler
    public void onArenaGameEnded(ArenaGameEndedEvent event) {
        MinigamesDTools.getInstance().getArenaAPI().restartArena(event.getArena().getName());
    }

    @EventHandler
    public void onInterract(PlayerInteractEvent event) {
        if(event.getItem() == null || event.getItem().getType() != Material.BOW) {
            return;
        }

        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(event.getPlayer());
        if(arena == null) {
            return;
        }

        if(arena.getState() == ArenaBase.STATE.COUNTDOWN || arena.getState() == ArenaBase.STATE.PAUSED) {
            ItemStack oldItem = event.getPlayer().getInventory().getItemInMainHand().clone();
            event.getPlayer().getInventory().setItemInMainHand(new ItemStack(Material.AIR));
            event.getPlayer().sendMessage("{match_hsnt_started_yet}");
            Bukkit.getScheduler().scheduleSyncDelayedTask(MinigamesDTools.getInstance(), () -> event.getPlayer().getInventory().setItemInMainHand(oldItem), 1L);
        }
    }

    @EventHandler
    public void onMove(PlayerMoveEvent event) {
        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(event.getPlayer());
        if(arena == null) {
            return;
        }

        if((arena.getGameRules().beforeFightDisableMoving && arena.getState() == ArenaBase.STATE.COUNTDOWN) || arena.getState() == ArenaBase.STATE.PAUSED) {
            if(event.getFrom().getX() != event.getTo().getX() || event.getFrom().getY() != event.getTo().getY() || event.getFrom().getZ() != event.getTo().getZ() ) {
                event.setCancelled(true);
            }
        }
    }

    /*
    * <PlayerDMG>
    * */
    @EventHandler
    public void onEntityDamageEvent(EntityDamageEvent event) {
        if(!(event.getEntity() instanceof Player)) {
            return;
        } else if(event instanceof EntityDamageByEntityEvent) {
            return;
        }

        Player player = (Player)event.getEntity();
        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(player);
        if(arena == null) {
            return;
        }

        if(arena.getState() != ArenaBase.STATE.EMPTY) {
            if((player.getHealth() - event.getFinalDamage()) <= 0D) {
                event.setDamage(0D);
                ArenaPlayerDeathLocalEvent arenaEvent = new ArenaPlayerDeathLocalEvent(player);
                arena.getEventAnnouncer().announce(arenaEvent);
                if(arenaEvent.isCancelled()) {
                    event.setCancelled(true);
                }
            } else {
                ArenaPlayerDamagedLocalEvent arenaEvent = new ArenaPlayerDamagedLocalEvent(player, event.getFinalDamage());
                arena.getEventAnnouncer().announce(arenaEvent);
                if(arenaEvent.isCancelled()) {
                    event.setCancelled(true);
                }
            }

        } else if(arena.getState() == ArenaBase.STATE.PAUSED && arena.getState() == ArenaBase.STATE.COUNTDOWN) {
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void onEntityDamageEvent(EntityDamageByEntityEvent event) {
        if(!(event.getEntity() instanceof Player)) {
            return;
        }

        Player player = (Player)event.getEntity();
        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(player);
        if(arena == null) {
            return;
        }

        if(arena.getState() != ArenaBase.STATE.EMPTY) {
            if(event.getDamager() instanceof Player) {
                if(arena.getTeamController().getPlayersRelative(player, (Player) event.getDamager()) == ArenaPlayersRelative.TEAMMATE) {
                    if(!arena.getTeamController().getTeamOf(player).friendlyFireAllowed()) {
                        event.setCancelled(true);
                        event.getDamager().sendMessage("{attack_teammate_msg}");
                        return;
                    }
                }
            }

            if((player.getHealth() - event.getFinalDamage()) <= 0.0D) {
                event.setDamage(0D);
                event.setCancelled(true);
                ArenaPlayerKilledLocalEvent arenaEvent = new ArenaPlayerKilledLocalEvent(player, event.getDamager());
                arena.getEventAnnouncer().announce(arenaEvent);
            } else {
                ArenaEntityDamagePlayerLocalEvent arenaEvent = new ArenaEntityDamagePlayerLocalEvent(event.getDamager(), player, event.getFinalDamage());
                arena.getEventAnnouncer().announce(arenaEvent);
                if(arenaEvent.isCancelled()) {
                    event.setCancelled(true);
                    event.setDamage(0D);
                }
            }
        } else if(arena.getState() == ArenaBase.STATE.PAUSED || arena.getState() == ArenaBase.STATE.COUNTDOWN) {
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void onPlayerDeathEvent(PlayerDeathEvent event) {
        if(event.getEntity() != null) {
            Player player = event.getEntity();
            ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(player);
            if(arena == null) {
                return;
            }

            ArenaPlayerDeathLocalEvent arenaEvent = new ArenaPlayerDeathLocalEvent(player);
            arena.getEventAnnouncer().announce(arenaEvent);
        }
    }

    /*
    * </playerDmg>
    * */

    @EventHandler
    public void onFoodLevelChangeEvent(FoodLevelChangeEvent event) {
        if(event.getEntity() instanceof Player) {
            Player player = (Player)event.getEntity();
            ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(player);
            if(arena != null && arena.getState() != ArenaBase.STATE.EMPTY && arena.getGameRules().hungerDisable) {
                event.setCancelled(true);
            }
        }
    }

    @EventHandler
    public void onBedjoin(PlayerBedEnterEvent event) {
        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(event.getPlayer());
        if(arena != null && arena.getState() != ArenaBase.STATE.PAUSED) {
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void onPlayerKick(PlayerKickEvent event) {
        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(event.getPlayer());
        if(arena != null && arena.getState() != ArenaBase.STATE.EMPTY) {
            ArenaPlayerLeaveLocalEvent arenaEvent = new ArenaPlayerLeaveLocalEvent(event.getPlayer());
            arena.getEventAnnouncer().announce(arenaEvent);
        }
    }


    @EventHandler
    public void onLogout(PlayerQuitEvent event) {
        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(event.getPlayer());
        if(arena != null && arena.getState() != ArenaBase.STATE.EMPTY) {
            ArenaPlayerLeaveLocalEvent arenaEvent = new ArenaPlayerLeaveLocalEvent(event.getPlayer());
            arena.getEventAnnouncer().announce(arenaEvent);
        }
    }

    /*
    @EventHandler
    public void onClickEvent(InventoryClickEvent event) {
        Player player = (Player)event.getWhoClicked();
        if(player == null) {
            return;
        }
        if(!MinigamesDTools.getInstance().getArenaAPI().getCacheManager().contains(player)) {

            if(event.getCurrentItem() == null) {
                return;
            } else if(!event.getCurrentItem().hasItemMeta()) {
                return;
            } else if(!event.getCurrentItem().getItemMeta().hasDisplayName()) {
                return;
            }

            if(event.getCurrentItem().getItemMeta().getDisplayName().equals("[arena]")) {
                event.getWhoClicked().sendMessage(MinigamesDTools.getPrefix() + ChatColor.RED + " С арены запрещено выносить вещи. Ваш предмет был заменён на землю.");
                event.getCurrentItem().setType(Material.DIRT);
                for(Enchantment enchantment : event.getCurrentItem().getEnchantments().keySet()) {
                    event.getCurrentItem().removeEnchantment(enchantment);
                }
                event.getCurrentItem().setItemMeta(null);
            }
        }
    }
    */

    @EventHandler
    public void onPlayerDrop(PlayerDropItemEvent event) {
        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(event.getPlayer());
        if(arena != null && !arena.getGameRules().playerCanItemDrop && arena.getState() != ArenaBase.STATE.EMPTY) {
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void onPlayerPickupDrop(EntityPickupItemEvent event) {
        if(!(event.getEntity() instanceof Player)) {
            return;
        }

        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf((Player) event.getEntity());
        if(arena != null && !arena.getGameRules().playerCanItemPickup && arena.getState() != ArenaBase.STATE.EMPTY) {
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void onPlayerDamage(PlayerItemDamageEvent event) {
        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(event.getPlayer());
        if(arena == null) {
            return;
        }

        ArenaBase.STATE arenaState = arena.getState();
        if(arenaState == null) {
            Debug.print(Debug.LEVEL.WARNING, "Detected invalid null state for arena[name:" + arena.getName() + "]");
        } else if(arenaState == ArenaBase.STATE.PAUSED || arenaState == ArenaBase.STATE.COUNTDOWN) {
            event.getPlayer().sendMessage("{damage_reject_while_cooldown_msg}");
            event.setCancelled(true);
        }
    }

    @EventHandler
    public void onTeleport(PlayerTeleportEvent event) {
        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf(event.getPlayer());
        if(arena != null && (arena.getState() == ArenaBase.STATE.COUNTDOWN || arena.getState() == ArenaBase.STATE.PAUSED)) {
            if(event.getFrom().distance(event.getTo()) > 1.0D) {
                event.setCancelled(true);
                event.getPlayer().sendMessage("{teleport_reject_msg}");
            }
        }
    }

    @EventHandler
    public void onRegain(EntityRegainHealthEvent event) {
        if(!(event.getEntity() instanceof Player)) return;

        ArenaBase arena = MinigamesDTools.getInstance().getArenaAPI().getArenaOf((Player) event.getEntity());
        if(arena != null && !arena.getGameRules().playerCanRegainHealth) {
            event.setCancelled(true);
        }
    }

    @EventHandler(
            priority = EventPriority.HIGHEST,
            ignoreCancelled = true
    )
    public void WeatherChangeEvent(WeatherChangeEvent event) {
        if (event.toWeatherState()) {
            event.setCancelled(true);
            event.getWorld().setWeatherDuration(0);
            event.getWorld().setThundering(false);
        }
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/listener/Events.kt
package com.gmail.borlandlp.minigamesdtools.listener

import com.gmail.borlandlp.minigamesdtools.MinigamesDTools
import com.gmail.borlandlp.minigamesdtools.arena.ArenaBase
import com.gmail.borlandlp.minigamesdtools.arena.ArenaPlayersRelative
import com.gmail.borlandlp.minigamesdtools.arena.localevent.*
import com.gmail.borlandlp.minigamesdtools.events.ArenaGameEndedEvent
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.entity.*
import org.bukkit.event.player.*
import org.bukkit.event.weather.WeatherChangeEvent
import org.bukkit.inventory.ItemStack

class Events : Listener {
    @EventHandler
    fun onArenaGameEnded(event: ArenaGameEndedEvent) {
        MinigamesDTools.instance!!.arenaAPI!!.restartArena(event.arena.name!!)
    }

    @EventHandler
    fun onInteract(event: PlayerInteractEvent) {
        if (event.item == null || event.item.type != Material.BOW) {
            return
        }

        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(event.player) ?: return
        if (arena.state == ArenaBase.STATE.COUNTDOWN || arena.state == ArenaBase.STATE.PAUSED) {
            val oldItem = event.player.inventory.itemInMainHand.clone()
            event.player.inventory.itemInMainHand = ItemStack(Material.AIR)
            event.player.sendMessage("{match_hsnt_started_yet}")
            Bukkit.getScheduler().scheduleSyncDelayedTask(
                MinigamesDTools.instance,
                { event.player.inventory.itemInMainHand = oldItem },
                1L
            )
        }
    }

    @EventHandler
    fun onMove(event: PlayerMoveEvent) {
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(event.player) ?: return
        if (arena.gameRules!!.beforeFightDisableMoving && arena.state == ArenaBase.STATE.COUNTDOWN || arena.state == ArenaBase.STATE.PAUSED) {
            if (event.from.x != event.to.x || event.from.y != event.to.y || event.from.z != event.to.z) {
                event.isCancelled = true
            }
        }
    }

    /*
    * <PlayerDMG>
    * */
    @EventHandler
    fun onEntityDamageEvent(event: EntityDamageEvent) {
        if (event.entity !is Player || event is EntityDamageByEntityEvent) {
            return
        }
        val player = event.entity as Player
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(player) ?: return
        if (arena.state != ArenaBase.STATE.EMPTY) {
            if (player.health - event.finalDamage <= 0.0) {
                event.damage = 0.0
                val arenaEvent =
                    ArenaPlayerDeathLocalEvent(
                        player
                    )
                arena.eventAnnouncer.announce(arenaEvent)
                if (arenaEvent.isCancelled) {
                    event.isCancelled = true
                }
            } else {
                val arenaEvent =
                    ArenaPlayerDamagedLocalEvent(
                        player,
                        event.finalDamage
                    )
                arena.eventAnnouncer.announce(arenaEvent)
                if (arenaEvent.isCancelled) {
                    event.isCancelled = true
                }
            }
        } else if (arena.state == ArenaBase.STATE.PAUSED && arena.state == ArenaBase.STATE.COUNTDOWN) {
            event.isCancelled = true
        }
    }

    @EventHandler
    fun onEntityDamageEvent(event: EntityDamageByEntityEvent) {
        if (event.entity !is Player) {
            return
        }
        val player = event.entity as Player
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(player) ?: return
        if (arena.state != ArenaBase.STATE.EMPTY) {
            if (event.damager is Player) {
                val isFriendFire = arena.teamController!!.getPlayersRelative(
                        player,
                        event.damager as Player
                    ) == ArenaPlayersRelative.TEAMMATE
                if (isFriendFire && !arena.teamController!!.getTeamOf(player)!!.isFriendlyFireAllowed) {
                    event.isCancelled = true
                    event.damager.sendMessage("{attack_teammate_msg}")
                    return
                }
            }

            if (player.health - event.finalDamage <= 0.0) {
                event.damage = 0.0
                event.isCancelled = true
                val arenaEvent =
                    ArenaPlayerKilledLocalEvent(
                        player,
                        event.damager
                    )
                arena.eventAnnouncer.announce(arenaEvent)
            } else {
                val arenaEvent =
                    ArenaEntityDamagePlayerLocalEvent(
                        event.damager,
                        player,
                        event.finalDamage
                    )
                arena.eventAnnouncer.announce(arenaEvent)
                if (arenaEvent.isCancelled) {
                    event.isCancelled = true
                    event.damage = 0.0
                }
            }
        } else if (arena.state == ArenaBase.STATE.PAUSED || arena.state == ArenaBase.STATE.COUNTDOWN) {
            event.isCancelled = true
        }
    }

    @EventHandler
    fun onPlayerDeathEvent(event: PlayerDeathEvent) {
        val player = event.entity ?: return
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(player) ?: return
        val arenaEvent =
            ArenaPlayerDeathLocalEvent(player)
        arena.eventAnnouncer.announce(arenaEvent)
    }

    /*
    * </playerDmg>
    * */

    @EventHandler
    fun onFoodLevelChangeEvent(event: FoodLevelChangeEvent) {
        if (event.entity !is Player) {
            return
        }

        val player = event.entity as Player
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(player) ?: return
        event.isCancelled = arena.state != ArenaBase.STATE.EMPTY && arena.gameRules!!.hungerDisable
    }

    @EventHandler
    fun onBedjoin(event: PlayerBedEnterEvent) {
        if (MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(event.player) == null) {
            return
        }
        event.isCancelled = true
    }

    @EventHandler
    fun onPlayerKick(event: PlayerKickEvent) {
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(event.player) ?: return
        if (arena.state != ArenaBase.STATE.EMPTY) {
            arena.eventAnnouncer.announce(
                ArenaPlayerLeaveLocalEvent(
                    event.player
                )
            )
        }
    }

    @EventHandler
    fun onLogout(event: PlayerQuitEvent) {
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(event.player) ?: return
        if (arena.state != ArenaBase.STATE.EMPTY) {
            arena.eventAnnouncer.announce(
                ArenaPlayerLeaveLocalEvent(
                    event.player
                )
            )
        }
    }

    /*
    @EventHandler
    public void onClickEvent(InventoryClickEvent event) {
        Player player = (Player)event.getWhoClicked();
        if(player == null) {
            return;
        }
        if(!MinigamesDTools.getInstance().getArenaAPI().getCacheManager().contains(player)) {

            if(event.getCurrentItem() == null) {
                return;
            } else if(!event.getCurrentItem().hasItemMeta()) {
                return;
            } else if(!event.getCurrentItem().getItemMeta().hasDisplayName()) {
                return;
            }

            if(event.getCurrentItem().getItemMeta().getDisplayName().equals("[arena]")) {
                event.getWhoClicked().sendMessage(MinigamesDTools.getPrefix() + ChatColor.RED + " С арены запрещено выносить вещи. Ваш предмет был заменён на землю.");
                event.getCurrentItem().setType(Material.DIRT);
                for(Enchantment enchantment : event.getCurrentItem().getEnchantments().keySet()) {
                    event.getCurrentItem().removeEnchantment(enchantment);
                }
                event.getCurrentItem().setItemMeta(null);
            }
        }
    }
    */
    @EventHandler
    fun onPlayerDrop(event: PlayerDropItemEvent) {
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(event.player) ?: return
        event.isCancelled = !arena.gameRules!!.playerCanItemDrop && arena.state != ArenaBase.STATE.EMPTY
    }

    @EventHandler
    fun onPlayerPickupDrop(event: EntityPickupItemEvent) {
        if (event.entity !is Player) {
            return
        }
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(event.entity as Player) ?: return
        event.isCancelled = !arena.gameRules!!.playerCanItemPickup && arena.state != ArenaBase.STATE.EMPTY
    }

    @EventHandler
    fun onPlayerDamage(event: PlayerItemDamageEvent) {
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(event.player) ?: return
        if (arena.state == ArenaBase.STATE.PAUSED || arena.state == ArenaBase.STATE.COUNTDOWN) {
            event.player.sendMessage("{damage_reject_while_cooldown_msg}")
            event.isCancelled = true
        }
    }

    @EventHandler
    fun onTeleport(event: PlayerTeleportEvent) {
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(event.player) ?: return
        if (arena.state == ArenaBase.STATE.COUNTDOWN || arena.state == ArenaBase.STATE.PAUSED) {
            if (event.from.distance(event.to) > 1.0) {
                event.isCancelled = true
                event.player.sendMessage("{teleport_reject_msg}")
            }
        }
    }

    @EventHandler
    fun onRegain(event: EntityRegainHealthEvent) {
        if (event.entity !is Player) return
        val arena = MinigamesDTools.instance!!.arenaAPI!!.getArenaOf(event.entity as Player) ?: return
        event.isCancelled = !arena.gameRules!!.playerCanRegainHealth
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onWeatherChangeEvent(event: WeatherChangeEvent) {
        if (event.toWeatherState()) {
            event.isCancelled = true
            event.world.weatherDuration = 0
            event.world.isThundering = false
        }
    }
}